src/helpers/proposed_network.rs

```diff
@@ -0,0 +1,198 @@
+}
+
+pub trait Ring: Sync {
+    /// Type of the channel that is used to send messages to other helpers
+    type Sink: futures::Sink<RingEvent, Error = NetworkEventError> + Send + Unpin + 'static;
+    type MessageStream: Stream<Item = RingEvent> + Send + Unpin + 'static;
+
+    /// Returns a sink that accepts data to be sent to other helper parties.
+    fn sink(&self) -> Self::Sink;
+
+    /// Returns a stream to receive messages that have arrived from other helpers. Note that
+    /// some implementations may panic if this method is called more than once.
+    fn recv_stream(&self) -> Self::MessageStream;
+}
+
+pub struct CreateQueryData {
+    callback: oneshot::Sender<QueryId>,
+}
+
+impl CreateQueryData {
+    #[must_use]
+    pub fn new(callback: oneshot::Sender<QueryId>) -> Self {
+        CreateQueryData { callback }
+    }
+    pub fn respond(self, query_id: QueryId) -> Result<(), NetworkEventError> {
+        self.callback
+            .send(query_id)
+            .map_err(|_| NetworkEventError::CallbackFailed {
+                event_name: "CreateQuery",
+                query_id,
+            })
+    }
+}
+
+pub struct PrepareQueryData {
+    pub query_id: QueryId,
+    callback: oneshot::Sender<()>,
+}
+
+impl PrepareQueryData {
+    #[must_use]
+    pub fn new(query_id: QueryId, callback: oneshot::Sender<()>) -> Self {
+        PrepareQueryData { query_id, callback }
+    }
+
+    pub fn respond(self) -> Result<(), NetworkEventError> {
+        let query_id = self.query_id;
+        self.callback
+            .send(())
+            .map_err(|_| NetworkEventError::CallbackFailed {
+                event_name: "PrepareQuery",
+                query_id,
+            })
+    }
+}
+
+pub struct StartMulData {
+    pub query_id: QueryId,
+    pub endpoints_positions: [Uri; 3],
+    pub field_type: String,
+    pub data_stream: ByteArrStream,
+    callback: oneshot::Sender<()>,
+}
+
+impl StartMulData {
+    pub fn new(
+        query_id: QueryId,
+        endpoints_positions: [Uri; 3],
+        field_type: String,
+        data_stream: ByteArrStream,
+        callback: oneshot::Sender<()>,
+    ) -> Self {
+        StartMulData {
+            query_id,
+            endpoints_positions,
+            field_type,
+            data_stream,
+            callback,
+        }
+    }
+
+    pub fn respond(self) -> Result<(), NetworkEventError> {
+        let query_id = self.query_id;
+        self.callback
+            .send(())
+            .map_err(|_| NetworkEventError::CallbackFailed {
+                event_name: "StartMul",
+                query_id,
+            })
+    }
+}
+
+pub struct MulData {
+    pub query_id: QueryId,
+    pub endpoints_positions: [Uri; 3],
+    pub endpoints_to_roles_and_stream: HashMap<Uri, (Role, ByteArrStream)>,
+}
+
+impl MulData {
+    pub fn new(
+        query_id: QueryId,
+        endpoints_positions: [Uri; 3],
+        endpoints_to_roles_and_stream: HashMap<Uri, (Role, ByteArrStream)>,
```

Comment:
why there is more than one stream? each helper including leader should receive only one stream to process

Reply:
this is 1 message that includes both streams intended for both other helpers. the network layer would break this up and send the respective stream to each other helper

Reply:
I just want to make sure we model for commands and events having different structure. For example, command to send data to the followers  may take two streams as its input, but event generated on each helper will contain only one stream.

Reply:
right. i see what you mean. i guess technically this is possible because the followers would receive a hashmap with exactly 1 entry (their own). that's...a bit clunky. i can rewrite it so that the command only has 1 stream, and the leader just has to send 2 of those commands

Reply:
> with exactly 1 entry (their own). that's...a bit clunky. 

100% agree. we don't have to use the same concept to represent command and events, they can be entirely separate and use different structure.

Reply:
i see. i actually might disagree -- i really like the conciseness of having a "command" just pipe through unchanged. Leader helper commands that a follower helper does something, and follower helper `respond`s to that command. it's simpler if you don't change the semantics of command in the middle

Reply:
that would work if sender creates the command exactly how it should be received by the recipient. In this particular case, QPL needs to send two different commands that carry over streams that intended for one recipient only. That works too

Reply:
Transport in this case is dumb and just delivers what it was told to deliver. No additional routing per command is supported